### PR TITLE
Fix issue #368 and add tests

### PR DESCRIPTION
**Pull request by dylanjcastillo**
Hey @EmilStenstrom @Fingel,

I got a working version of a fix for issue #368. There was a quick fix (but @EmilStenstrom you might not be its biggest fan 😅). The hardest part was getting tests that replicated the issue, but I now have a version that seems to achieve that.

@Fingel your research was very useful to understand where the issue originated. Thank you very much.

When Django [initializes](https://docs.djangoproject.com/en/5.0/ref/applications/#initialization-process) the application registry, for each application in `INSTALLED_APPS` Django imports the root package of the application.

For `django-components`, this means the `ready` method is called which then calls `autodiscover`, as pointed by Fingel. `autodiscover` looks for all the locations of components and then executes those files.

The issue occurs during the execution of the component files found in `autodiscover`, specifically in `spec.loader.exec_module(module)`. Due to differences in the way the modules are imported I couldn't find a way to verify if two components were the same except for checking the file it was coming from + the name of the class.

I found two ways to do that:

1. `sys.modules[cls.__module__].__file__`: I think this should always work, but I'm not really sure, I have some gaps on how some of the `spec` and `module` stuff works. So, I'd rather go for a less standard approach, but one that I feel more confident about.
2. `inspect.getfile(cls)`: This is the method I proposed previously. Based on my tests it works well, and even though it might seem a bit hacky, other projects such as spaCy seem to use it for [similar purposes](https://github.com/explosion/spaCy/blob/a493981163002d0cd2409950512eeeccb6fa4690/spacy/util.py#L1143).

I propose we use `inspect.getfile()` fix for now. I think it's worth it, given the use cases it unblocks. OTOH, for future releases (including some of the stuff we were discussing in the HTMX-enabled components thread), it might be worthwhile to rethink the whole registration process, given that Components are becoming more complex now.

Also, I found an unrelated issue in `autodiscover` that, depending on the path resolution, resulted in duplicated directories when running `Loader.get_dirs`. This wasn't causing this issue, but I fixed it as well (see changes to `template_loader.py`. Maybe it'd speed up a bit the initialization process.

Finally, after some research, I realized that there's [no guarantee](https://docs.djangoproject.com/en/5.0/ref/applications/#django.apps.AppConfig.ready) that `ready` will be called only once when initializing the application. So, `autodiscover` and thus the registry of components must be robust to be running them multiple times. At first, I thought that we could prevent `autodiscover` from running twice, but I don't think that makes a lot of sense based on this new information.

TL;DR:

1. I propose we use `inspect.getfile()` as it fixes the issue for now.
2. I got tests that replicate the issue.
3. I also found a smaller bug I fixed in `template_loader.py`
4. For future releases it might be worthwhile to rethink the whole registration process

### CODE REVIEW

**Comment by EmilStenstrom** [comment]
@dylanjcastillo Thanks again!

@Fingel Just released [0.34.1](https://pypi.org/project/django-components/0.34.1/), can you verify it works?

**Comment by Fingel** [comment]
Nice work, I had also considered using `inspect.getfile()` in the hash function. Kinda crazy how hard it is to uniquely identify a class like that.
